### ISSUE TITLE
Declarative token creation

### ISSUE THREAD

**Issue by CRASH-Tech** [argoproj/argo-cd#9884]
# Summary

Declarative create static tokens

# Motivation

I'm using helm for create app projects and roles. Then I create token by hands, which call argo-cd from CI. It's a pain and not declarative.

# Proposal

Keep static tokens in secret or AppProject CRD.

**Comment by crenshaw-dev** [comment]
Tokens are signed with the API server's key, so they must be generated at runtime. You can always declaratively store the tokens in Secrets using SealedSecrets or something similar.

**Comment by todaywasawesome** [comment]
@CRASH-Tech I agree with @crenshaw-dev here. Argo CD is setup so you can use your own secrets solution for this which is a better solution than trying to get everyone to use one specific one or to reinvent the wheel with something new.

**Comment by CRASH-Tech** [comment]
> Tokens are signed with the API server's key, so they must be generated at runtime. You can always declaratively store the tokens in Secrets using SealedSecrets or something similar.

How I can do it?

**Comment by crenshaw-dev** [comment]
@CRASH-Tech I assume you're currently generating a token with something like `argocd proj role create-token`. After generating it, you'd just store it in any mechanism that allows you to generate a Secret (such as in a SealedSecret if you have that controller installed, or in Vault if you have a Vault controller installed, etc.). Once it's in a Secret in Kubernetes, you can reference it wherever you need it.

**Comment by crenshaw-dev** [comment]
If your CI is already capable of creating AppProjects and Applications, then it is effectively admin. What is the need for the project token / where is that token eventually used?

**Comment by CRASH-Tech** [comment]
> If your CI is already capable of creating AppProjects and Applications, then it is effectively admin. What is the need for the project token / where is that token eventually used?

I can't manual sync or refresh over crd's. Only over cli.

**Comment by crenshaw-dev** [comment]
Ah! So you want to create a project token with only Kubernetes access, not Argo CD access. You could 

1) add `accounts.some-username: apiKey, login` to argocd-cm
1) add `accounts.some-username.password: 'some-bcrypted-password'` to argocd-secret
1) after creating an AppProject, log in to Argo CD with the CLI using the newly-created account
1) use the CLI to create a project token
1) use the project token to do your syncs/refreshes
1) delete the account?

It's a bit janky, but I think it would accomplish what you're trying to do.

**Comment by iamyeka** [comment]
I got the same case. I need to specify the token when doing helm installation. After the installation, i can simply use the token in my program to create application or other things.

**Comment by ArjonBu** [comment]
@crenshaw-dev after the deletion of server.secretkey, then the token will not work anymore.

**Comment by closetool** [comment]
@CRASH-Tech 
Here my solution
1. add `serverSignature` to argocd-secret which used to generate jwt token.
2. generate a token
3. add `accounts.{accountName}.tokens` to argocd-secret, takes claims `id`, `iat` from token, and set it as `accounts.{accountName}.tokens`'s value, like this `[{"id":"bd05a278-2c86-4047-9bd8-ef64209dffa5","iat":1673236464},{"id":"865d160e-288f-4edb-bb82-02278a5fa070","iat":1673255112}]`
after that, you can use your token freely

**Comment by jakirpatel** [comment]
+1 btw is there any document to understand how to statically declare the token ?

**Comment by quintonn** [comment]
@kilosonc could you please elaborate on step 2, generate token?

I'm trying to add a token as part of a cdk script, and this ticket is the only reference I have found with instructions for adding a password and token into the instance.
The password code works, and I want to also add and test the token component.

Also, what library would generate the serverSignature and should it be added to the values.yaml config too?

**Comment by crenshaw-dev** [comment]
@jackmtrayport I'm not sure the docs say you can create the tokens declaratively. As a matter of fact, I think they say the opposite:

```yaml
    # NOTE: JWT tokens can only be generated by the API server and the token is not persisted
    # anywhere by Argo CD. It can be prematurely revoked by removing the entry from this list.
```

I opened https://github.com/argoproj/argo-cd/pull/11736 and then realized I didn't have enough context to understand the change, so I closed it.

**Comment by crenshaw-dev** [comment]
I think it would be easy enough to create a small controller to generate tokens and save them to Secrets. The controller could either call the Argo CD API Server or just directly use the server.secretKey and re-implement JWT generation.

I'm trying to decide whether I think that's a good feature to add to Argo CD itself. Feels more like an argoproj-labs thing, since JWT creation/maintenance doesn't fit neatly into any existing Argo CD component.

**Comment by crenshaw-dev** [comment]
> The example yaml containing the jwtTokens: entry does apply and creates a token that you can see in the Argo UI

My guess is that it's meant to go the other way around: when you create a token, Argo CD adds an entry to `jwtTokens` which you can then delete to disable the token. But if you add an entry yourself, no token is actually ever created.

> Either way though I think the docs should be updated to make it clear that while you can specify jwtTokens in a yaml that you kubectl apply to your cluster, you shouldn't do it because it's absolutely pointless (and that section of the example yaml should be removed).

Yep, the docs should be updated. I'm not sure we should remove the YAML entirely, since it does communicate useful functionality (disabling a previously-created token). But it should be reworded to indicate that the field doesn't actually _create_ any tokens.

**Comment by zeusal** [comment]
About this, where `do you store the token generated through the API?

**Comment by leoandrea7** [comment]
Any news?

**Comment by ArshamTeymouri** [comment]
Any updates?
The issue was opened almost two years ago without any resolution so far. There are many people reporting the same problem in different discussion threads, PRs and issues but still no workaround for it. Generating a jwt token without being able to specify its value is absolutely pointless as it doesn't allow people to use it in their pipelines confidently.
refs:
- https://github.com/argoproj/argo-cd/discussions/12486
- https://github.com/argoproj/argo-cd/issues/3213
- https://github.com/argoproj/argo-cd/issues/14250
- https://github.com/argoproj/argo-cd/pull/19572

@jannfis do you have an idea why this is not implemented yet?

**Comment by BulatSaif** [comment]
Did anyone find a workaround?
